### PR TITLE
Onboarding Brand Design Update: Add address bar position screen

### PR DESCRIPTION
**Pull request by mikescamell**
Task/Issue URL: https://app.asana.com/1/137249556945/project/1207908166761516/task/1212699261164257

### Description
Implements the address bar position selection screen for the onboarding brand design update. This is the second dialog step in the redesigned onboarding flow, allowing users to choose between top, bottom, or split address bar positions.

Key changes:
- Address bar position dialog with animated title typing, bobbing Dax, and option toggles
- Step indicator view showing progress through onboarding dialogs
- Bottom wing Lottie animation with dismissal logic
- Animated bubble arrow during comparison chart transition
- Background images for address bar and comparison chart screens
- Multiple bug fixes for rotation, tablet layout, and animation edge cases

### Steps to test this PR

[Designs](https://www.figma.com/design/YPE94Xkcrk2uqiF2l4VmSv/Onboarding--2026-?node-id=16507-46861&m=dev)

Apply patch

```
Subject: [PATCH] onboarding
---
Index: app/src/main/java/com/duckduckgo/app/onboarding/ui/OnboardingViewModel.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/app/src/main/java/com/duckduckgo/app/onboarding/ui/OnboardingViewModel.kt b/app/src/main/java/com/duckduckgo/app/onboarding/ui/OnboardingViewModel.kt
--- a/app/src/main/java/com/duckduckgo/app/onboarding/ui/OnboardingViewModel.kt	(revision 06d747bcc29beada0a444df876b05f55f0dbb4d5)
+++ b/app/src/main/java/com/duckduckgo/app/onboarding/ui/OnboardingViewModel.kt	(date 1775628931668)
@@ -45,7 +45,7 @@
     val viewState = _viewState.asStateFlow()
 
     fun initializePages() {
-        pageLayoutManager.buildPageBlueprints()
+        pageLayoutManager.buildBrandDesignUpdatePageBlueprints()
     }
 
     fun pageCount(): Int {
@@ -70,7 +70,7 @@
     }
 
     fun initializeOnboardingSkipper() {
-        if (!appBuildConfig.canSkipOnboarding) return
+        return
 
         // delay showing skip button until privacy config downloaded
         viewModelScope.launch {

```

_Comparison chart screen_
- [x] Open a fresh install and go through onboarding to the address bar position screen
- [x] Verify comparison chart transition animates correctly
- [x] Verify check icon staggered bounce animation plays
- [x] Rotate and verify state is restored
- [x] Press "Choose Your Browser"
- [x] Ensure wing animates out and you transition to the Address bar position screen
- [x] See next steps

_Address bar position screen_
- [x] Verify title typing animation plays on entry
- [x] Verify bobbing Dax animation slides in and bobs
- [x] Toggle between top, bottom, and split address bar options
- [x] Rotate the device and verify the dialog restores correctly with the selected option preserved
- [x] Verify step indicator shows correct step count
- [x] Primary CTA does nothing (will work for when Search/AI input is added)

### UI changes
See [here](https://app.asana.com/1/137249556945/task/1213895377190156?focus=true) for demoes

---

> [!NOTE]
> **Medium Risk**
> Moderate risk: adds a new onboarding dialog step with multiple coordinated animations, state restoration, and tablet-specific inset/layout handling that could regress the onboarding flow on different form factors.
> 
> **Overview**
> Adds an **address bar position selection** step to the brand-design-update onboarding flow, including new UI for top/bottom/split options with themed radio-button tinting, new toolbar preview drawables (light/dark, selected/unselected), and a bobbing Dax animation.
> 
> Extends onboarding transitions to support a new `OnboardingBackgroundStep.AddressBar` background, updates `BrandDesignUpdateWelcomePage` to drive the new dialog’s animations and option selection wiring (including split-option gating and rotation-safe “show without animation” state), and refines tablet handling via a shared `DeviceInfo.isTablet()` helper and tablet-specific insets.
> 
> Improves `OnboardingStepIndicatorView` by pinning the text label width to prevent layout jitter as steps change.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 93470c30d2c00a4425339b5c37a033689b2a693f. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by mikescamell** [comment]
* **#8223** <a href="https://app.graphite.com/github/pr/duckduckgo/Android/8223?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8216** <a href="https://app.graphite.com/github/pr/duckduckgo/Android/8216?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/duckduckgo/Android/8216?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `develop`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Comment by mikescamell** [comment]
bugbot run

File: app/src/main/java/com/duckduckgo/app/onboarding/ui/page/BrandDesignUpdateWelcomePage.kt [hunk 1]
```diff
@@ -689,7 +754,96 @@ class BrandDesignUpdateWelcomePage : OnboardingPageFragment(R.layout.content_onb
...
+                                    addListener(object : AnimatorListenerAdapter() {
+                                        override fun onAnimationEnd(animation: Animator) {
+                                            isAnimating = false
+                                        }
+                                    })
+                                    start()
+                                }
+                            }
+                        }
+                    })
```
[the remark this answers is not in the document]
Comment (mikescamell) [reply app/src/main/java/com/duckduckgo/app/onboarding/ui/page/BrandDesignUpdateWelcomePage.kt]: Fixed in a6a1001. Transition and listener now tracked in changeBoundsTransition/changeBoundsTransitionListener for cleanup in onDestroyView, and guard changed from `!isAdded` to `view == null`.

File: app/src/main/res/drawable/mobile_toolbar_top_unselected_brand_design_update_light.xml [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        android:fillColor="#000000"
+        android:pathData="M50 98v2H10v-2h40zm8-8V10c0-4.42-3.58-8-8-8H10C5.72 2 2.23 5.36 2.01 9.59L2 10v80c0 4.42 3.58 8 8 8v2c-5.35 0-9.72-4.2-9.99-9.49L0 90V10C0 4.48 4.48 0 10 0h40l0.51 0.01C55.8 0.28 60 4.65 60 10v80l-0.01 0.51c-0.26 5.12-4.36 9.22-9.48 9.48L50 100v-2c4.42 0 8-3.58 8-8z" />
+    <path
+        android:fillAlpha="0.06"
+        android:fillColor="#000000"
+        android:pathData="M2 10c0-4.42 3.58-8 8-8h40c4.42 0 8 3.58 8 8v8H2v-8z" />
+    <path
+        android:fillAlpha="0.36"
+        android:fillColor="#000000"
+        android:pathData="M2 18h56v2H2v-2zm6-8c0-2.21 1.79-4 4-4h24c2.21 0 4 1.79 4 4s-1.79 4-4 4H12c-2.21 0-4-1.79-4-4zm36 0c0-2.21 1.79-4 4-4s4 1.79 4 4-1.79 4-4 4-4-1.79-4-4z" />
```
[the remark this answers is not in the document]
Comment (mikescamell) [reply app/src/main/res/drawable/mobile_toolbar_top_unselected_brand_design_update_light.xml:37]: Fixed in a6a1001. Aligned top unselected light drawable to use `#242323` with alpha `0.4`, matching bottom and split variants.

**Comment by mikescamell** [comment]
bugbot run

File: app/src/main/java/com/duckduckgo/app/onboarding/ui/page/BrandDesignUpdateWelcomePage.kt [hunk 3]
```diff
@@ -689,7 +754,99 @@ class BrandDesignUpdateWelcomePage : OnboardingPageFragment(R.layout.content_onb
...
+
+                    binding.daxDialogCta.root.translationZ = 1f.toPx()
+                    TransitionManager.beginDelayedTransition(binding.daxDialogCta.root as ViewGroup, transition)
+
+                    binding.daxDialogCta.comparisonChartContent.root.isVisible = false
+                    binding.daxDialogCta.addressBarContent.root.isVisible = true
+
+                    binding.daxDialogCta.primaryCta.text = getString(R.string.preOnboardingAddressBarOkButton)
+                    binding.daxDialogCta.primaryCta.setOnClickListener { viewModel.onPrimaryCtaClicked() }
+                    binding.daxDialogCta.primaryCta.alpha = 0f
```
[the remark this answers is not in the document]
Comment (mikescamell) [reply app/src/main/java/com/duckduckgo/app/onboarding/ui/page/BrandDesignUpdateWelcomePage.kt:852]: Fixed in 0bfe881. Moved `primaryCta.setOnClickListener` into the `onAnimationEnd` callback of `addressBarFadeInAnimatorSet`, matching the pattern used by COMPARISON_CHART. The button is no longer tappable while invisible during the transition.

File: app/src/main/java/com/duckduckgo/app/onboarding/ui/page/BrandDesignUpdateWelcomePage.kt [hunk 4]
```diff
@@ -847,8 +1005,68 @@ class BrandDesignUpdateWelcomePage : OnboardingPageFragment(R.layout.content_onb
...
+                if (binding.daxDialogCta.addressBarContent.root.isVisible) {
+                    setAddressBarPositionOptions(selectedAddressBarPosition, showSplitOption)
+                    return
+                }
+
                 binding.bottomWingAnimation.isVisible = false
-                // TODO
+
+                binding.logoAnimation.alpha = 0f
+                binding.welcomeTitle.alpha = 0f
```
[the remark this answers is not in the document]
Comment (mikescamell) [reply app/src/main/java/com/duckduckgo/app/onboarding/ui/page/BrandDesignUpdateWelcomePage.kt]: Fixed in 0bfe881. Removed the duplicate `logoAnimation.alpha = 0f` and `welcomeTitle.alpha = 0f` assignments — already set before the early-return guard.

**Comment by mikescamell** [comment]
bugbot run

File: app/src/main/java/com/duckduckgo/app/onboarding/ui/page/BrandDesignUpdateWelcomePage.kt [hunk 5]
```diff
@@ -134,9 +141,55 @@ class BrandDesignUpdateWelcomePage : OnboardingPageFragment(R.layout.content_onb
...
+            bottomOmnibarToggleImage.setImageResource(omnibarToggleImageRes(OmnibarType.SINGLE_BOTTOM, selectedOption, isLightMode))
+            bottomOmnibarToggleCheck.isChecked = selectedOption == OmnibarType.SINGLE_BOTTOM
+            splitOmnibarToggleImage.setImageResource(omnibarToggleImageRes(OmnibarType.SPLIT, selectedOption, isLightMode))
+            splitOmnibarToggleCheck.isChecked = selectedOption == OmnibarType.SPLIT
+
+            splitOmnibarContainer.isVisible = showSplitOption
+
+            topOmnibarToggleImage.setOnClickListener {
+                viewModel.onAddressBarPositionOptionSelected(OmnibarType.SINGLE_TOP)
+                setAddressBarPositionOptions(OmnibarType.SINGLE_TOP, showSplitOption)
```
Comment (LukasPaczos) [inline app/src/main/java/com/duckduckgo/app/onboarding/ui/page/BrandDesignUpdateWelcomePage.kt]: nit: might be cleaner to rely on the state in the view model instead of changing the option directly on button click, just in case the state is ever modified outside of the button interactions

File: app/src/main/java/com/duckduckgo/app/onboarding/ui/page/BrandDesignUpdateWelcomePage.kt [hunk 6]
```diff
@@ -689,7 +754,99 @@ class BrandDesignUpdateWelcomePage : OnboardingPageFragment(R.layout.content_onb
 
                 ADDRESS_BAR_POSITION -> {
                     dismissBottomWingAnimation()
-                    // TODO
+                    val isTablet = resources.configuration.smallestScreenWidthDp >= 600
```
Comment (LukasPaczos) [inline app/src/main/java/com/duckduckgo/app/onboarding/ui/page/BrandDesignUpdateWelcomePage.kt]:
```suggestion
                    val isTablet = deviceInfo.formFactor() == DeviceInfo.FormFactor.TABLET
```
same in all the other places

File: app/src/main/java/com/duckduckgo/app/onboarding/ui/page/BrandDesignUpdateWelcomePage.kt [hunk 7]
```diff
@@ -847,8 +1005,65 @@ class BrandDesignUpdateWelcomePage : OnboardingPageFragment(R.layout.content_onb
...
+                binding.daxDialogCta.stepIndicator.alpha = 1f
+                binding.daxDialogCta.stepIndicator.setSteps(viewModel.getMaxPageCount(), 2)
+                binding.daxDialogCta.primaryCta.alpha = 1f
+                binding.daxDialogCta.primaryCta.text = getString(R.string.preOnboardingAddressBarOkButton)
+                binding.daxDialogCta.primaryCta.setOnClickListener { viewModel.onPrimaryCtaClicked() }
+
+                binding.daxDialogCta.root.isVisible = true
+                binding.daxDialogCta.daxCtaContainer.alpha = 1f
+
+                binding.bobbingDaxAnimation?.apply {
```
Comment (LukasPaczos) [inline app/src/main/java/com/duckduckgo/app/onboarding/ui/page/BrandDesignUpdateWelcomePage.kt]:
```suggestion
                binding.bobbingDaxAnimation.apply {
```

File: app/src/main/java/com/duckduckgo/app/onboarding/ui/page/BrandDesignUpdateWelcomePage.kt [hunk 8]
```diff
@@ -421,6 +480,12 @@ class BrandDesignUpdateWelcomePage : OnboardingPageFragment(R.layout.content_onb
         changeBoundsTransitionListener = null
         changeBoundsTransition = null
         binding.daxDialogCta.comparisonChartContent.comparisonChartTitle.cancelAnimation()
+        binding.daxDialogCta.addressBarContent.addressBarTitle.cancelAnimation()
+        addressBarFadeInAnimatorSet?.cancel()
+        addressBarFadeInAnimatorSet = null
+        bobbingDaxAnimator?.cancel()
+        bobbingDaxAnimator = null
+        binding.bobbingDaxAnimation?.cancelAnimation()
```
Comment (LukasPaczos) [inline app/src/main/java/com/duckduckgo/app/onboarding/ui/page/BrandDesignUpdateWelcomePage.kt]:
```suggestion
        binding.bobbingDaxAnimation.cancelAnimation()
```

File: app/src/main/res/layout/include_brand_design_address_bar_position.xml [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+            app:layout_constraintWidth_max="160dp">
+
+            <ImageView
+                android:id="@+id/topOmnibarToggleImage"
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content"
+                android:contentDescription="@string/preOnboardingAddressBarPositionTop"
+                android:foreground="@drawable/selectable_rounded_ripple"
+                android:scaleType="fitCenter"
+                android:src="@drawable/mobile_toolbar_top_selected_light" />
```
Comment (LukasPaczos) [inline app/src/main/res/layout/include_brand_design_address_bar_position.xml]:
```suggestion
                tools:src="@drawable/mobile_toolbar_top_selected_brand_design_update_light" />
```
nit, same below, it gets immediately overwritten anyway

**Review by LukasPaczos — COMMENTED** [review]
Works well 👍 Some comments below - I think the most important is to address the radio button touch area, rest we can address here or in follow ups, or not at all.

1. The actual radio button is not clickable, only the pictogram is.

https://github.com/user-attachments/assets/51dd2ff5-7196-4f9d-9514-231ef84a26b0

2. The background doesn't go all the way to the bottom edge on Samsung Tab S9+. Not an issue on Pixel 9 Pro.
Image

3. `configureDaxCta` function is already over 300 LOC, and it's not even all of it yet. Similar pattern of huge switch-cases repeats across the classes. I understand you're following a previously established pattern, but this project would be a great opportunity to improve that code organization. It's not something to address in this PR but maybe worth considering before shipping all of the steps?

**Comment by mikescamell** [comment]
> Works well 👍 Some comments below - I think the most important is to address the radio button touch area, rest we can address here or in follow ups, or not at all.
> 
> 1. The actual radio button is not clickable, only the pictogram is.
> 
>  Screen_recording_20260409_082104.mp4 
> 2. The background doesn't go all the way to the bottom edge on Samsung Tab S9+. Not an issue on Pixel 9 Pro.
> 
> Image
> 3. `configureDaxCta` function is already over 300 LOC, and it's not even all of it yet. Similar pattern of huge switch-cases repeats across the classes. I understand you're following a previously established pattern, but this project would be a great opportunity to improve that code organization. It's not something to address in this PR but maybe worth considering before shipping all of the steps?

1. Yeah good point, I need to figure out how we can handle the click ripples, because it looked odd previously where both the image and the radio button would ripple individually

2. That is odd, I'll look into this but first time I'm seeing a device not respect edgetoedge for the new onboarding

3. Agreed, I wanted to ask Claude to refactor this after the final dialog PR to get something a little nicer 👍

**Comment by mikescamell** [comment]
> > Works well 👍 Some comments below - I think the most important is to address the radio button touch area, rest we can address here or in follow ups, or not at all.
> > 
> > 1. The actual radio button is not clickable, only the pictogram is.
> > 
> > Screen_recording_20260409_082104.mp4
> > 2. The background doesn't go all the way to the bottom edge on Samsung Tab S9+. Not an issue on Pixel 9 Pro.
> > Image
> > 3. `configureDaxCta` function is already over 300 LOC, and it's not even all of it yet. Similar pattern of huge switch-cases repeats across the classes. I understand you're following a previously established pattern, but this project would be a great opportunity to improve that code organization. It's not something to address in this PR but maybe worth considering before shipping all of the steps?
> 
> 1. Yeah good point, I need to figure out how we can handle the click ripples, because it looked odd previously where both the image and the radio button would ripple individually
> 2. That is odd, I'll look into this but first time I'm seeing a device not respect edgetoedge for the new onboarding
> 3. Agreed, I wanted to ask Claude to refactor this after the final dialog PR to get something a little nicer 👍



> > Works well 👍 Some comments below - I think the most important is to address the radio button touch area, rest we can address here or in follow ups, or not at all.
> > 
> > 1. The actual radio button is not clickable, only the pictogram is.
> > 
> > Screen_recording_20260409_082104.mp4
> > 2. The background doesn't go all the way to the bottom edge on Samsung Tab S9+. Not an issue on Pixel 9 Pro.
> > Image
> > 3. `configureDaxCta` function is already over 300 LOC, and it's not even all of it yet. Similar pattern of huge switch-cases repeats across the classes. I understand you're following a previously established pattern, but this project would be a great opportunity to improve that code organization. It's not something to address in this PR but maybe worth considering before shipping all of the steps?
> 
> 1. Yeah good point, I need to figure out how we can handle the click ripples, because it looked odd previously where both the image and the radio button would ripple individually
> 2. That is odd, I'll look into this but first time I'm seeing a device not respect edgetoedge for the new onboarding
> 3. Agreed, I wanted to ask Claude to refactor this after the final dialog PR to get something a little nicer 👍

2. Seems to be a Samsung tablet issue, and it's applying a scrim. The onboarding is still edgetoedge and this doesn't seem to happen on phones as evidenced from this screenshot. I'm going to leave this as is. I do not think it's detrimental to the designs, and we get a similar effect if someone is using 3 button nav instead of gestures regardless.

image

File: app/src/main/java/com/duckduckgo/app/onboarding/ui/page/BrandDesignUpdateWelcomePage.kt [hunk 10]
```diff
@@ -689,7 +751,98 @@ class BrandDesignUpdateWelcomePage : OnboardingPageFragment(R.layout.content_onb
...
+                    binding.daxDialogCta.root.translationZ = 1f.toPx()
+                    TransitionManager.beginDelayedTransition(binding.daxDialogCta.root as ViewGroup, transition)
+
+                    binding.daxDialogCta.comparisonChartContent.root.isVisible = false
+                    binding.daxDialogCta.addressBarContent.root.isVisible = true
+
+                    binding.daxDialogCta.primaryCta.text = getString(R.string.preOnboardingAddressBarOkButton)
+                    binding.daxDialogCta.primaryCta.alpha = 0f
+
+                    setAddressBarPositionOptions(OmnibarType.SINGLE_TOP, showSplitOption)
```
Comment (LukasPaczos) [inline app/src/main/java/com/duckduckgo/app/onboarding/ui/page/BrandDesignUpdateWelcomePage.kt]: This should use value from the state, not hardcode to top.

File: app/src/main/java/com/duckduckgo/app/onboarding/ui/page/BrandDesignUpdateWelcomePage.kt [hunk 11]
```diff
@@ -702,6 +855,7 @@ class BrandDesignUpdateWelcomePage : OnboardingPageFragment(R.layout.content_onb
     private fun showDialogWithoutAnimation(
         onboardingDialogType: PreOnboardingDialogType,
         showSplitOption: Boolean = false,
+        selectedAddressBarPosition: OmnibarType = OmnibarType.SINGLE_TOP,
```
Comment (LukasPaczos) [inline app/src/main/java/com/duckduckgo/app/onboarding/ui/page/BrandDesignUpdateWelcomePage.kt]:
```suggestion
        selectedAddressBarPosition: OmnibarType,
```
Do we need the default value? It should always be available from state.

**Review by LukasPaczos — COMMENTED** [review]
The full container ripple doesn't look great:

https://github.com/user-attachments/assets/e2c86c5a-ef95-4668-8ecc-64085716a29a

I think we should do the same we do in the appearance and AI settings screens, so that only the pictogram ripples:

https://github.com/user-attachments/assets/ea86293f-9128-4ffb-8a04-5e291d05ca2c

**Review by LukasPaczos — APPROVED** [review]
LGTM!

File: app/src/main/res/layout/include_brand_design_address_bar_position.xml [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+                android:text="@string/preOnboardingAddressBarPositionTop"
+                android:textAppearance="@style/Typography.DuckDuckGo.Onboarding.Label" />
+
+            <com.duckduckgo.common.ui.view.button.RadioButton
+                android:id="@+id/topOmnibarToggleCheck"
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content"
+                android:layout_marginTop="6dp"
+                android:background="@null"
+                android:checked="true"
```
[the remark this answers is not in the document]
Comment (mikescamell) [reply app/src/main/res/layout/include_brand_design_address_bar_position.xml]: Removed the hardcoded `android:checked="true"`. The ViewModel default (`SINGLE_TOP`) is the single source of truth and `setAddressBarPositionOptions()` always sets checked state before the container becomes visible.

File: app/src/main/java/com/duckduckgo/app/onboarding/ui/page/BrandDesignUpdateWelcomePage.kt [hunk 13]
```diff
@@ -847,8 +1009,64 @@ class BrandDesignUpdateWelcomePage : OnboardingPageFragment(R.layout.content_onb
...
+                    if (deviceInfo.isTablet()) {
+                        verticalBias = 0.5f
+                        bottomToTop = binding.bobbingDaxAnimation.id
+                        bottomToBottom = ConstraintLayout.LayoutParams.UNSET
+                    } else {
+                        verticalBias = 0f
+                        bottomToTop = ConstraintLayout.LayoutParams.UNSET
+                        bottomToBottom = ConstraintLayout.LayoutParams.PARENT_ID
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (mikescamell) [reply app/src/main/java/com/duckduckgo/app/onboarding/ui/page/BrandDesignUpdateWelcomePage.kt:1036]: Fixed. Switched from `.layoutParams...apply { }` to `updateLayoutParams<ConstraintLayout.LayoutParams> { }` to ensure `requestLayout()` is called, matching the COMPARISON_CHART case.
